File: Aula04/Go/Atv03/main.go
package main

import (
	"fmt"
)

type Pessoa struct {
	Nome     string
	Idade    int
	Endereco Endereco
}

type Endereco struct {
	Rua    string
	Cidade string
	Estado string
	CEP    string
}

type Empresa struct {
	Nome         string
	CNPJ         string
	Funcionarios []Pessoa
}

func (Pessoa *Pessoa) Adicionar_endereco(Endereco *Endereco) {
	Pessoa.Endereco = *Endereco
}

func (Pessoa *Pessoa) Mostrar_informacoes() {
	fmt.Println("Nome:", Pessoa.Nome, ", Idade:", Pessoa.Idade)
	Pessoa.Endereco.Mostrar_endereco()
}

func (Endereco *Endereco) Mostrar_endereco() {
	fmt.Println("Rua:", Endereco.Rua, ", Cidade:", Endereco.Cidade, ", Estado:", Endereco.Estado, ", CEP:", Endereco.CEP)
}

func (Empresa *Empresa) Constratar_funcionario(funcionario Pessoa) {
	Empresa.Funcionarios = append(Empresa.Funcionarios, funcionario)
}

func (Empresa *Empresa) Listar_funcionarios() {
	fmt.Println("Funcionários da Empresa", Empresa.Nome)
	for _, funcionario := range Empresa.Funcionarios {
		fmt.Println(funcionario.Nome)
	}
}

func main() {

	var endereco1 Endereco = Endereco{"Av. Principal", "Cidade A", "EStado X", "12345-678"}
	var pessoa1 Pessoa = Pessoa{"João", 30, Endereco{}}
	pessoa1.Adicionar_endereco(&endereco1)
	pessoa1.Mostrar_informacoes()

	var endereco2 Endereco = Endereco{"Rua Secundária", "Cidade B", "EStado Y", "98765-432"}
	var pessoa2 Pessoa = Pessoa{"Maria", 25, Endereco{}}
	pessoa2.Adicionar_endereco(&endereco2)
	pessoa2.Mostrar_informacoes()

	var empresa = Empresa{"ABC", "12345678901234", nil}
	empresa.Constratar_funcionario(pessoa1)
	empresa.Constratar_funcionario(pessoa2)

	empresa.Listar_funcionarios()

}

File: Aula04/Python/Atv03/main.py
class Pessoa:
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade
        self.endereco = None
    
    def adicionar_endereco(self, endereco):
        self.endereco = endereco

    def mostrar_informacoes(self):
        print(f"Nome: {self.nome}, Idade: {self.idade}")
        if self.endereco:
            print("Endereço: ")
            self.endereco.mostrar_endereco()
        else:
            print("Endereço não disponível")
    
class Endereco:
    def __init__(self, rua, cidade, estado, cep):
        self.rua = rua
        self.cidade = cidade
        self.estado = estado
        self.cep = cep
    
    def mostrar_endereco(self):
        print(f"Rua: {self.rua}, Cidade: {self.cidade}, Estado: {self.estado}, CEP: {self.cep}")

class Empresa:
    def __init__(self, nome, cnpj):
        self.nome = nome
        self.cnpj = cnpj
        self.funcionarios = []

    def constratar_funcionario(self, funcionario):
        self.funcionarios.append(funcionario)

    def listar_funcionarios(self):
        print(f"Funcionários da empresa {self.nome}:")
        for funcionario in self.funcionarios:
            print(funcionario.nome)

endereco1 = Endereco("Av. Principal", "Cidade A", "EStado X", "12345-678")
pessoa1 = Pessoa("João", 30)
pessoa1.adicionar_endereco(endereco1)


endereco2 = Endereco("Rua Secundária", "Cidade B", "EStado Y", "98765-432")
pessoa2 = Pessoa("Maria", 25)
pessoa2.adicionar_endereco(endereco2)

empresa = Empresa("Empresa ABC", "12345678901234")
empresa.constratar_funcionario(pessoa1)
empresa.constratar_funcionario(pessoa2)

empresa.listar_funcionarios()
